src/Compilers/CSharp/Portable/Binder/NameofBinder.cs

```diff
@@ -2,19 +2,51 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
+    /// <summary>
+    /// If a proper method named "nameof" exists in the outer scopes, this binder does nothing.
+    /// Otherwise, it relaxes the instance-vs-static requirement for top-level member access expressions
+    /// and when inside an attribute on a method it adds type parameters from the target of that attribute.
+    /// To do so, it works together with <see cref="ContextualAttributeBinder"/>.
+    /// 
+    /// For other attributes (on types, type parameters or paramaeters) we use a WithTypeParameterBinder directly
+    /// in the binder chain and some filtering (<see cref="LookupOptions.MustNotBeMethodTypeParameter"/>) to keep
+    /// pre-existing behavior.
+    /// </summary>
     internal sealed class NameofBinder : Binder
     {
         private readonly SyntaxNode _nameofArgument;
+        private readonly WithTypeParametersBinder? _nextWhenNameofOperatorInAttribute;
```

Comment:
>_nextWhenNameofOperatorInAttribute [](http://example.com/codeflow?start=51&length=34)

The name feels very confusing. This is never the Next binder. The chain of Next binders is unconditional and can be walked by anyone. This is an optional "underlying" ```WithTypeParametersBinder``` to which we delegate some lookup requests. We are not exposing this binder to anyone, this is just an implementation detail. Consider renaming.